### ISSUE TITLE
Implement multi-level dynamic types

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#2544]
In the dynamic types specification, we have established that Starcounter will auto-create "one extra level" of types, for every type declaration explicitly specified by the developer. This is not yet implemented.

This model

```C#
class Car {
  [Type] CarModel Model;
}
class CarModel {}
```

currently result in **one** new type: a type named "Car" (of .NET type ```CarModel```).

The current (logical) tuple of this type will look approx like:

Key | NET Type | Name | TypeInherits |Type
----- | ------------- | -------- | --------------- | -------
1 | CarModel | "Car" | null | null

The right thing for **next** is that it should instead result in **two** new types, with tuples looking like

Key | NET Type | Name | TypeInherits |Type
----- | ------------- | -------- | --------------- | -------
1 | CarModel | "Car" | null | 2
2 | [Unknown] | "CarModel" | null | null

("[Unknown]" is an implementation detail, but must support ```IObjectView``` and ```IDbTuple```; in the first implementation, it will most likely be represented by the metadata ```RawView```)

**Comment by per-samuelsson** [comment]
Just to be more specific, these levels can be further extended explicitly by the developer, as shown below. The important things Starcounter will need to assure is to always add one "extra" level, and to assure the type-references are correctly specified.

```C#
class Car {
  [Type] CarModel Model;
}
class CarModel {
  [Type] CarModelType Type;
}
class CarModelType {}
```

Key | NET Type | Name | TypeInherits |Type
----- | ------------- | -------- | --------------- | -------
1 | CarModel | "Car" | null | 2
2 | CarModelType | "CarModel" | null | 3
3 | [Unknown] | "CarModelType" | null | null

**Comment by per-samuelsson** [comment]
I have spent today further investigating and looking at how we should get this to work in the code host binding process. I now have a better understanding of what we need; all above examples are not complete (they all miss one layer).

I'll update here again once I have been able to craft a better diagram.

**Comment by per-samuelsson** [comment]
After even further studies, I'm now leaning towards using the **Entity table** (note: making no dependency on the ```Entity``` class here) for "first level" types, and use the ```RawView``` as the terminating ("right-most") type, as is illustrated by the examples below.

#### Case studies
Legend:
HostType**(**Key, Name, Type, Inherits**)**
HostTuple**[**Key, Type**]**

##### Case 1
```C#
// The simple Car - one class is two new
// types; with one, we can instantiate new
// cars, with the other, we can instantiate
// new car models.
[Database] class Car1 {}
```
* BIND(Car1):
  * Entity(1, "Car1", 2, null)
  * RawView(2, "Car1Type", null, null);
* new Car1():
  * Car1[3, 1];

**Comment by per-samuelsson** [comment]
I still haven't met any theoretical obstacles with the new implementation suggested above, so I'll go with that. 

One question I'm not 100% sure of is if I can use the ```RawView``` as the "terminating type", since it belongs to the metadata types and those are very strict in their implementation. It might fail on the IS-operator and the ```TypeBinding``` to use that, but I will at least make an attempt with that first.

**Comment by k-rus** [comment]
> Entity(1, "Car1", 2, null)
> RawView(2, "Car1Type", null, null);

A minor comment:
I guess from the implementation point of view the type of type should be created earlier, i.e. `Car1Type` should have `Key=1` and `Car1` - `Key=2`.



> One question I'm not 100% sure of is if I can use the RawView as the "terminating type", since it belongs to the metadata types and those are very strict in their implementation. It might fail on the IS-operator and the TypeBinding to use that, but I will at least make an attempt with that first.

Why don't you use `Entity` table instead for the type of type (terminator)? The semantic of `RawView` is to describe hard types corresponding, e.g., to CLR classes. The query processor then uses it to get to know about all hard types and uses dynamic type objects to get to know about names of soft types. Thus it is OK to use `RawView` if no new instances are created, but if new instances of `RawView` have to be created I expect some problems might appear.

> Case 3
> 
> // Here, things get interesting; with two user defined
> // classes, four types are introduced. And an important
> // distinction here is that Car3 specifies a user defined
> // layout/class for it's type; with this, the usage of the
> // entity table is no longer in play.
> [Database] class Car3 {[Type] Car4;}
> [Database] class Car4 {}
> BIND(Car3, Car4):
> CarModel(1, "Car3", 2, null);
> RawView(2, "Car3Type", null, null);
> Entity(3, "Car4", 4, null)
> RawView(4, "Car4Type", null, null);
> new Car3():
> Car3[5, 1];

I am wonder why the bound instances do not store type relationship from `Car3` to `Car4`?

**Comment by per-samuelsson** [comment]
Thanks for input, @k-rus!

> A minor comment:
I guess from the implementation point of view the type of type should be created earlier, i.e. Car1Type should have Key=1 and Car1 - Key=2.

Yes. As I explained before, I consider how we assign keys to be an implementation detail. It could have been 23587234 and 99999829 respectively, and in any order.

> Thus it is OK to use RawView if no new instances are created, but if new instances of RawView have to be created I expect some problems might appear.

Yeah, I never intended to create any new ```RawView``` instances, because that would have failed for sure. I just thought I would use the already existing ones to indicate termination. I was afraid something could fail just if I changed the ```RawView``` class, like adding a (computed) property and/or methods (since assertions are so rigid).

> Why don't you use Entity table instead for the type of type (terminator)?

I considered that, and I still do (because of above reasoning of "problematic" use of ```RawView```). My reason NOT to use Entity, was that ```RawView``` already exist, and in discussions we've had, we have said that "for types, we already have instances created, describing everything - we can use those" (meaning metadata types). But then again, I actually might end up using entity table also for terminaing type-types, if it's easier. Things will change anyway in next-next.

> I am wonder why the bound instances do not store type relationship from Car3 to Car4?

It does. The car instance reference the CarModel. There is no *inherit* relationship between the two, hence no such reference is part of the type model. (With that being said, there might be flaws still, since this is still very much work in progress :smile: ).

**Comment by k-rus** [comment]
> It does. The car instance reference the CarModel. There is no inherit relationship between the two, hence no such reference is part of the type model. (With that being said, there might be flaws still, since this is still very much work in progress  ).

I see that `CarModel` references key 2, i.e., `Car3Type`. Should `CarModel` reference key 3 instead? Another question: is `Car3Type` instance needed?



> Yeah, I never intended to create any new RawView instances, because that would have failed for sure. I just thought I would use the already existing ones to indicate termination. I was afraid something could fail just if I changed the RawView class, like adding a (computed) property and/or methods (since assertions are so rigid).

Thank you for the clarification!

**Comment by per-samuelsson** [comment]
> Another question: is Car3Type instance needed?

Yes. Because you always need a type to be able to instantiate something. And if you want to instantiate a new *type* of Car3, you need an instance to do it from.

As an example, to create a new Car3, you would do 

```C#
var type = Db.TypeOf<Car>();
var car = type.Create()
``` 

The ```car``` will have IsType=false and Type = [type]. 

To instead create a **specialization of the type**, I need the "type of the type", hence:

```C#
var type = Db.TypeOf<Car3>();
var subtype = type.Type.Create();
``` 

Now, ```subtype``` will have IsType=true and TypeInherits = [type].

**Comment by k-rus** [comment]
> Another question: is Car3Type instance needed?
> Yes. Because you always need a type to be able to instantiate something. And if you want to instantiate a new type of Car3, you need an instance to do it from.
> 
> As an example, to create a new Car3, you would do
> 
> var type = Db.TypeOf<Car>();
> var car = type.Create()
> The car will have IsType=false and Type = [type].
> 
> To instead create a specialization of the type, I need the "type of the type", hence:
> 
> var type = Db.TypeOf<Car3>();
> var subtype = type.Type.Create();
> Now, subtype will have IsType=true and TypeInherits = [type].

Why is `Car4` not used instead of `Car3Type`? We know from the class definitions that type of `Car3` is `Car4`.

**Comment by per-samuelsson** [comment]
> Why is Car4 not used instead of Car3Type?

The ```Car4``` **class** (a.k.a host type) is used, but the Car3 **type**. If you would reduce the set of types by one (as you suggest), how would you create **a)** instances of Car4 and **b)** specializations of Car4?

**Comment by per-samuelsson** [comment]
In **case 1**, we have established that

```C#
[Database] class Car1 {}
```

produce 2 new type instance in the database. We all agree on this, correct?

With that, the following would neccessarily produce the double, i.e 4, types:

```C#
[Database] class Car1 {}
[Database] class Car2 {}
```

Again, correct?

Now, if we change this by specifying a reference from ```Car1``` to ```Car2```, like

```C#
[Database] class Car1 {
  [Type] public Car2 Type;
}
[Database] class Car2 {}
```

why would it produce fewer types?

**Comment by per-samuelsson** [comment]
@k-rus,

Can I ask another couple of minutes of your help? Can you please revise the **Case 3**-study, and model it the way you expected it to be (like we discussed in the meeting)?

Please do this below, by modifying this very comment (just use the "Edit comment" button), so we get the formatting right too. Would be very much appreciated!

Legend:
HostType**(**Key, Name, Type, Inherits**)**
HostTuple**[**Key, Type**]**

##### Case 3 - revisited by @k-rus
```C#
[Database] class Car3 {[Type] Car4;}
[Database] class Car4 {}
```
* BIND(Car3, Car4):
  * Car4(1, "Car3", 2, null);
  * RawView(2, "Car3Type", null, null);
  * Entity(3, "Car4", 4, null)
  * RawView(4, "Car4Type", null, null);
* new Car3():
  * Car3[5, 1];

**Comment by k-rus** [comment]
I updated the above example, however it was not a simple task, since it was updated at the same time ;)

**Comment by per-samuelsson** [comment]
> however it was not a simple task, since it was updated at the same time ;)

Sorry! I did it like 10 secons after I wrote, since I missed to include the original. I didn't expect you to be that fast!

> I updated the above example

Thanks!

**Comment by per-samuelsson** [comment]
This frustrates me like crazy. I can still not formulate what is wrong with either approach. But neither explain what is more "right" with the other. I refuse to start any real implementation of this until a "formula" is fully clear on a correct approach, that includes a clear explaination of why things must be modelled one, and only one, way.

**Comment by k-rus** [comment]
I think it might be helpful to look into more complex example:
```c#
[Database]
public class CarModel { };
[Database]
public class Car {
    [Type] public CarModel Model;
}
public class Opel : Car { };
```
This example extends case 3 with inheritance on `Car`, i.e., `Car3`.
According to my understanding I expect the following type instances in the database after the declaration:

* BIND(Opel, Car, CarModel)
  * Entity(1, "CarModel", 2, null)
  * RawView(2, "auto_CarModelType", null, null)
  * CarModel(3, "Car", 1, null)
  * CarModel(4, "Opel", 5, 3)
  * RawView(5, "auto_CarModelSubType", null, 1) // May be the host type should Entity
* c = new Car()
  * Car[6, 3]
* o = new Opel()
  * Opel[7,4]

Therefore: 
```c#
o.Type.Name == "Opel"
o.Type.Type.Name == "auto_CarModelSubType"
o.Type.Inherits.Name == "Car"
o.Type.Type.Inherits.Name == "CarModel"
o.Type.Inherits.Type.Name == "CarModel"
```
Basically it is important to not have a paradox, i.e., if I go against inheritance and then against instantiation, I should get the same result if I go first against instantiation and then against inheritance:
```c#
o.Type.Inherits.Type == o.Type.Type.Inherits
```

**Comment by per-samuelsson** [comment]
Thanks for prooviding more of your thoughts, I'm still doing studies of this and I will now consider further your example.

Generally, I think we should switch to use Entity rather than RawView for terminating/auto-created types, based on previous reasoning. But of course it does not affect the actual problem we are discussing here.

**Comment by k-rus** [comment]
> Generally, I think we should switch to use Entity rather than RawView for terminating/auto-created types, based on previous reasoning. But of course it does not affect the actual problem we are discussing here.

I agree. It might help at this moment.

**Comment by per-samuelsson** [comment]
I can see the logic in your last example, and I don't see any immediate "wrongs" in it.

However, why did previous ```((Car3)c).Type.Type.Name == "Car4"``` and now ```o.Type.Type.Name == "auto_CarModelSubType"```? In the first statement, we ended up with the user defined "Car4" type name, in the second with an auto-generated type name. Isn't the model in this respect quite identical (minus that ```Opel``` is a specialization of  ```Car```)?

**Comment by k-rus** [comment]
> In the first statement, we ended up with the user defined "Car4" type name, in the second with an auto-generated type name. Isn't the model in this respect quite identical (minus that Opel is a specialization of Car)?

Noб the model is not identical. Every specialization on left side should be reflected on types inheritance. Thus type of Opel is auto-generated, since user has not created a specialization of CarModel. Basically my `auto_CarModelSubType` should be seen as `OpelType` or `OpelModel`.

**Comment by per-samuelsson** [comment]
> Thus type of Opel is auto-generated, since user has not created a specialization of CarModel. Basically my auto_CarModelSubType should be seen as OpelType or OpelModel.

Yeah, I guess that makes sense, and after all, auto_CarModelSubType still IS a CarModel, becuase of how the inhertance hierarchy is specified. I saw that now. But thanks for giving it more clarification.

**Comment by per-samuelsson** [comment]
After a weekend of hell, I think I now have been able to finally establish some kind of "formula" for the above models here. I really hope I can't find any holes in this now, since its what I will base my implementation on.

It's based on @k-rus's observations, which I think is right. Here it goes:

Any defined .NET **class** defines a **table** and a **type**. The *name* of the type is the class name. The *type* (i.e type *object*) of the type is either a system-defined, auto-generated type, *or* the type of the class specified as the **type class** of the current class. The **host type** (i.e. .NET class in the .NET binding) is either decided by the system (working model: Entity), or the class **type class**.

Hierarchies will solve themselves based on the above.

Pretty simple actually, but hard (at least for me) to reach still. I just cross my fingers now there is no glitch I have missed. :smile:

**Comment by per-samuelsson** [comment]
I have made an implementation now, that I'm testing. I have not done anything regarding migration, though, so if you change type-hierarchy dependant declarations, you will get an error.

I'm implementing regression test for @k-rus model from above. They seem to work.

However, this model - part of my previous dynamic types modelling regression test - currently fail, and I'm investigating why:

```C#
public class Something : Entity {
  [Type] public SomethingKind Kind;
}
public class SomethingKind : Something {}
```

**Comment by per-samuelsson** [comment]
After yet another day of working with this, there is still some flaw in the implementation that shows up when I do further and more complex multi-level models that include more levels defined than 3 (i.e. types to types to types).

I right now trying to take a step back (again) to straighten out any possibly questionmark.

**Comment by per-samuelsson** [comment]
Take 2:

**Why must there neccesarily be two type objects for one defined class?**
Consider the following model:

```C#
class Car {}
```

It defines a table: *Car*. And a type (object): *"Car"*. 

When *type* "Car" is instantiated, a new object/tuple is created in the Car table. This object/tuple reference as its type "Car". Hence, ```typeobject("Car").Create()``` inserts a new object in the Car table. With its (implicit) type reference set to ```typeobject("Car")```.

Legend:
* **Table**(ID, Name, Type, Inherits, **Instantiates**)
* any value *?* is still not defined
* any table name like **-table-** defines that the instance is **not** a type (i.e. ```IsType=false```)
* The name of table "Type" is not used in the implementation

Lets take a look at what we've got so far:

* BIND(Car)
  * **Type**(1, "Car", ?, ?, **Car**)
* new Car()
  * **-Car-**(2, null, 1, null, null)

An alternative to use ```new Car()``` is to start from the type object itself. Invoking ```Create()``` on it should produce the same result.

* Db.TypeOf("Car").Create()
  * **-Car-**(101, null, 1, null, null)

This all looks like we expect it to. 

Now, why do we need another type? Answer: to *specialize* the "Car" type. Hence, the full implementation and effect of binding the simple Car class definition is:

* BIND(Car)
  * **Type**(1, "Car", 2, null, **Car**)
  * **Type**(2, "Car__autoType", null, null, **Type**(1))

From this, we can see that traversing one step to the right from the original "Car" type, we end up with another type, *"Car__autoType"*. This is the type to use to create specializations of "Car". A specialization has a IS-relationship defined by the *Inherits* property; hence, you'll see that any new "Car" specialization created relates to the former with the same property.

* Db.TypeOf("Car").Type.Create()
  * **Type**(3, null, 2, 1, **Car**)

**Comment by per-samuelsson** [comment]
Lets continue.

```C#
class Opel : Car {}
```

Here, class Opel extends class Car. Hence, it defines a new table: *Opel*. And a type (object): *"Opel"*. 

With class inheritance introduced, our underlying type model will now extend like this:

* BIND(Car, Opel)
  * **Type**(1, "Car", 2, null, **Car**)
  * **Type**(2, "Car__autoType", null, null, **Type**(1))
  * **Type**(3, "Opel", 4, 1, **Opel**)
  * **Type**(4, "Opel__autoType", null, 2, **Type**(2))

**Comment by per-samuelsson** [comment]
I think I **might** finally found the flaw with @k-rus's suggested model, and why rationalizing away a level like that doesn't work - there is no way to create a Car-specialization.

```C#
var carModel = Db.TypeOf("CarModel").Create();
var carModelSpecialization = Db.TypeOf("CarModel").Type.Create();
var car = Db.TypeOf("Car").Create();
var carSpecialization = Db.TypeOf("Car").Type.Create();  // This doesn't work
```

What doesn't work with the final statement (and, IMO, proves that the suggested model is wrong)  is that ```Db.TypeOf("Car").Type.Create()```must absolutely create a new type that extends the "Car" type, i.e Inherits=3. It can't possibly do so, since the type we issue Create() on is the type of CarModel (that has no reference whatsoever to Car)

**Comment by k-rus** [comment]
> var carModel = Db.TypeOf("CarModel").Create();
> var carModelSpecialization = Db.TypeOf("CarModel").Type.Create();
> var car = Db.TypeOf("Car").Create();
> var carSpecialization = Db.TypeOf("Car").Type.Create();

Here it will be created (first from class definitions then from the runtime code):

* BIND(Car, CarModel)
  * **Type**(1, "CarModel", 2, null)
  * **Type**(2, "auto_CarModelType", null, null)
  * **Type**(3, "Car", 2, null)
* var carModel = Db.TypeOf("CarModel").Create();
  * ** Type**(4, "auto_CarModelSpecialization", null, 1)
  * **Type**(5, "unknown_line1", 4, 3)
* var carModelSpecialization = Db.TypeOf("CarModel").Type.Create();
  * **Type**(6, "auto_CarModelTypeSpecialization", null, 2)
  * **Type**(7, "uknnown_line2", 6, 1)
* var car = Db.TypeOf("Car").Create();
  * **Car**(8, "unknown_line3", 3, null)
* var carSpecialization = Db.TypeOf("Car").Type.Create();
  * **Type**(9, "auto_CarModelSpecialization2", null, 1);
  * **Type**(10, "unknown_line4", 9, 3)

What I wrote is according my understanding of the conceptual picture demonstrated by @Starcounter-Jack.

Basically the runtime code will create two inheritances from `Car` and `CarModel`. We can check with @Starcounter-Jack if I understood the algorithm correctly.

**Comment by k-rus** [comment]
In theory creation of each `Car` instance by `Db.TypeOf("Car").Create` or `new Car()` should create new child of `Type(3, "Car", 2, null)`. This will double number of objects in the database, thus we skip to create them.

**Comment by per-samuelsson** [comment]
> In theory creation of each Car instance by Db.TypeOf("Car").Create or new Car() should create new child of Type(3, "Car", 2, null). This will double number of objects in the database, thus we skip to create them.

Yes, this is known, but lets not bring this into the discussion; its out of scope since he have all agreed on optimizing it away.

I'll get back to your previous example.

**Comment by per-samuelsson** [comment]
> I'll get back to your previous example

Actually, I have quite a hard time understanding it.

* it doesn't match your previous suggestion
* I see not a single **CarModel** (host type) instance
* I see some new "uknnown_xxx" naming convention that has not been part of any previous discussion

**Comment by k-rus** [comment]
I see know that with this algorithm it is difficult to get the three objects from binding of `Car` and `CarModel`.

**Comment by k-rus** [comment]
> I see some new "uknnown_xxx" naming convention that has not been part of any previous discussion

What should be the names of the explicitly created instances if you haven't set the name value?

**Comment by k-rus** [comment]
> I see not a single CarModel (host type) instance

There are only instances, which inherits from `CarModel`.

**Comment by per-samuelsson** [comment]
> What should be the names of the explicitly created instances if you haven't set the name value?

We have used **null** so far. I thought we could stick to that.

> There are only instances, which inherits from CarModel.

So you mean the following code should raise a cast exception?

```C#
var model = (CarModel) Db.TypeOf("CarModel").Create();
```

**Comment by k-rus** [comment]
> What should be the names of the explicitly created instances if you haven't set the name value?
> We have used null so far. I thought we could stick to that.

Sorry, I forgot about this. So my unknowns are nulls.



> So you mean the following code should raise a cast exception?
> 
> var model = (CarModel) Db.TypeOf("CarModel").Create();

No, it will not raise exception, since `CarModel` is more general class.

**Comment by per-samuelsson** [comment]
> No, it will not raise exception, since CarModel is more general class.

OK, so you just didn't properly denote it using the previous legend? That made me confused. Especially since you used **Car** (correctly) on object with ID 8. But not **CarModel** any single time.

**Comment by k-rus** [comment]
> OK, so you just didn't properly denote it using the previous legend? That made me confused. Especially since you used Car (correctly) on object with ID 8. But not CarModel any single time.

`CarModel`, i.e., Key 1 is inherited by Keys 4 and 9.

You are right my mistake is that Key 8 should have type Key 1. Fixed.

**Comment by k-rus** [comment]
> You are right my mistake is that Key 8 should have type Key 1. Fixed.

Key 3 should have type Key 1.

**Comment by per-samuelsson** [comment]
> Key 3 should have type Key 1.

Thats still not what I asked for. :smile: You don't have any single instance that has *host type* / *table* **CarModel**.

Look at your instance with Key 8. It specifies **Car** as the host type / table. All others specify **Type**. If my suggested cast from above should work, there must be at least one of the instaces that have host type **CarModel**.

**Comment by k-rus** [comment]
> Thats still not what I asked for.  You don't have any single instance that has host type / table CarModel.
> 
> Look at your instance with Key 8. It specifies Car as the host type / table. All others specify Type. If my suggested cast from above should work, there must be at least one of the instaces that have host type CarModel.

I don't understand meaning of **Type**. Each **CarModel** instance is also **Type**. Thus I am confused. 

I fixed that CLR type is `CarModel` if the instance has a type inheriting from `CarModel`. I hope I haven't made more mistakes.

**Comment by per-samuelsson** [comment]
> I don't understand meaning of Type.

Ah, sorry, then my updated legend was not clear enough I guess. It's basically the substitute of **Entity**. I wanted to get away from **Entity** since it doesn't neccessarily have to be either class or table Entity in the implementation. So instead the legend stated

> The name of table "Type" is not used in the implementation

and

* that **CarModel** indicated that table and host type CarModel was used, **and** that this was a type (i.e. ```IsType=true```),
* that **-Car-** was using table/host type Car, but that this was NOT a type, and finally,
* that **Type** denoted a type (```IsType=true```) where the table/host type used was an implementation detail.

> I fixed that CLR type is CarModel if the instance has a type inheriting from CarModel.

Super. Its much clearer. Thanks.

**Comment by k-rus** [comment]
Thanks for the clarification!
I read the legend yesterday and didn't read again today morning. Some details I missed.


> that -Car- was using table/host type Car, but that this was NOT a type, and finally,

I fixed this too.

**Comment by per-samuelsson** [comment]
> I fixed this too.

Thank you. Now it looks awesome! :+1: 

So back to my actual point - even with your now extended model, I still can't see how the following sequence can possibly work:

* var carSpecialization = Db.TypeOf("Car").Type.Create();
  * **Type**(9, "auto_CarModelSpecialization2", null, 1);
  * **CarModel**(10, null, 9, 3)

Lets take it step by step:

```C#
var t = Db.TypeOf("Car");
Console.Write(t.Name + " " + t.ID);
```

must write "Car 3", right?

Then I reach for the type of that type

```C#
var typeType = t.Type;
Console.Write(typeType.Name + " " + typeType.ID);
```

and we have now reached the auto-generated "CarModel", correct? Hence, the above must print "CarModel 1", correct?

Then, if I do ```typeType.Create()```, what information is available on the instance level of this type? I would say

* it knows what table to insert into ("CarModel")
* it knows what host type to use (class CarModel)
* it knows that the created instance should be referencing itself via the "Type" property (or, if we make the implementation more complex like your last suggestion, that we should add two objects - one that derive, and then one that points to the derived one).

But *how can it know that inheritance should point back to OID(3)*? This is the flaw IMO, and what invalidates the previous suggested model (as well as the new one).

The only reasonable answer I can think of is that you suggest that OID(1) has a reference back to OID(3), thereby forcing it to be used solely to create new types based on Car(i.e. OID(3)). But then another flaw appears: we can not create an instance on the same *level* as Car, i.e. a CarModel who inherits not Car, but CarModel.

**Comment by k-rus** [comment]
I think I don't understand the semantic of `TypeOf`. Thus my model might be wrong and needs to be specialized. Can you explain what operator `TypeOf`does?

**Comment by per-samuelsson** [comment]
> Can you explain what operator TypeOfdoes?

It fetch the *type object* of the named class.

**Comment by k-rus** [comment]
> Can you explain what operator TypeOfdoes?
> It fetch the type object of the named class.

There are two type objects are related to class. Am I correct that `TypeOf` returns a type object, which has the given name?

Then 

> var t = Db.TypeOf("Car");
> Console.Write(t.Name + " " + t.ID);
> must write "Car 3", right?

is correct.

> Then I reach for the type of that type
> 
> var typeType = t.Type;
> Console.Write(typeType.Name + " " + typeType.ID);
> and we have now reached the auto-generated "CarModel", correct? Hence, the above must print "CarModel 1", correct?

this is also correct.



> if I do typeType.Create(), what information is available on the instance level of this type?

According to @Starcounter-Jack magic, this is known:

> But how can it know that inheritance should point back to OID(3)? This is the flaw IMO, and what invalidates the previous suggested model (as well as the new one).

Only one type object can be instantiated for type type object. Thus only Object 3 has type Object 1 and Object 1 can have reference back to Object 3. Therefore it is known that `typeType.Create()` will create an object, which inherits from t, i.e., `Object 3`.

**Comment by per-samuelsson** [comment]
> There are two type objects are related to class. Am I correct that TypeOf returns a type object, which has the given name?

Yes. But only one is accessible using the class, or class name, as in ```Db.TypeOf<Car>()``` or ```Db.TypeOf("Car")```. The reach the other object, i.e. the type-type, you need to traverse the result, i.e using the ```Type``` (or corresponding) property.

I pretty satisfied with how I present it in this https://github.com/Starcounter/Starcounter/issues/2544#issuecomment-73901888.

> According to @Starcounter-Jack magic, this is known:

I don't think there is magic involved. I think the implementation details have not been fully worked out, thats all. And thats what we are trying to do here.

> Only one type object can be instantiated for type type object. Thus only Object 3 has type Object 1 and Object 1 can have reference back to Object 3. Therefore it is known that typeType.Create() will create an object, which inherits from t, i.e., Object 3.

This seems very much like a paradox. I doubt that is the intended feature. And it really goes against the previous discussion where we discissed ```new CarModel()``` that included @Starcounter-Jack.

With that suggestion, it also would not be possible to create any other CarModel instances as soon as one other class defines it as its type (i.e. using ```[Type]```). But as soon as such a declaration was removed, instantiation would again be allowed. I very much doubt that is correct.

**Comment by per-samuelsson** [comment]
Continuing exploring from this https://github.com/Starcounter/Starcounter/issues/2544#issuecomment-73905012. Let's take one step back, and look at another thing - we'll remove the ```Opel``` type, and introduce a ```CarModel```, which we'll define as the type class of ```Car```.

As short reptetition first:

We have seen that declaration

```C#
class Car {}
```

binds to the type system as

* BIND(Car)
  * **Type**(1, "Car", 2, null, **Car**)
  * **Type**(2, "Car__autoType", null, null, **Type**(1))

We have seen that it's possible to *instantiate* the Car type using either the **new** operator, or the ```Create()``` factory method. Both produce the same thing: instances of type Car.

```C#
var car = new Car();
var anotherCar = Db.TypeOf<Car>().Create();
```

Both cars will have their *type* reference pointing to KEY(1).

We have also seen that it is possible to *specialize*, or *extend*, the Car type. This is done using the type of the type, like this:

```C#
var hotRod = Db.TypeOf<Car>().Type.Create();
hotRod.Name = "HotRod";
```

Here, the new HotRod extend the Car type; hence, its **inherits** reference is set to KEY(1). We can create new hot rod instances by using ```hotRod.Create()```.

Now, lets extend the code model like this:

```C#
class CarModel {}
class Car { [Type] CarModel Type; }
```

**Comment by per-samuelsson** [comment]
Continuing from the [previous comment](https://github.com/Starcounter/Starcounter/issues/2544#issuecomment-74218466).

(move the code snippet here)

**Ramblings** (highlingting the very essence of the uncertainty being discussed)
The biggest confusion in the implementation is the semantics of the ```[Type]``` specifier. What exactly does it do?

1. It makes the autogenerated type of the class declaring it use the **data type** of the reference as the **host type**. Other than that, the type model is equivalent to if it would not be used.
2. It makes the auto-generated type of the declaring class use the type object of the referenced type as it's *type reference*.
3. 2. It makes the auto-generated type of the declaring class use the type object of the referenced type as it's *inherits reference*.
4. It omits generating a new type object as a whole - it uses the one defined by the referencing class.

All 2-4 alternatives would affect how the referenced class (i.e. CarModel in this context) would be bound when isolated, i.e. they rely on the fact that the referenced type class is dedicated as the type class of the current, declaring class only.

The effect of 2-4 is that ```Db.TypeOf<CarModel>().Create()``` will create a Car. And if so, **new CarModel()** must neccesarily create a specialization of CarModel.

**Comment by per-samuelsson** [comment]
In the upcoming discussion we have planned, we should revise the ```TypeOf``` method.

Currently, ```Db.TypeOf<Car>``` and ```Db.TypeOf("Car")``` return the same object. That might actually be a bit confusing. In the former, it's clear that we want to go from a **class** to it's corresponding **type object**. In the later - do we want to get the type of the type named "Car", or the "Car" type itself?

It might be a good idea to deprecate the overload that takes a name, and only support going from the type class. This also correspond better to the ```typeof``` operator in C#. We could then additionally introduce static method ```GetType(string name)```, so that ```Db.TypeOf<Car>``` and ```Db.GetType("Car")``` return the same object.

**Comment by k-rus** [comment]
> var hotRod = Db.TypeOf<Car>().Type.Create();
> hotRod.Name = "HotRod";

The result of this type operation is that the type of `Car` has name *HotRod*.



> class CarModel {}
> class Car { [Type] CarModel Type; }

This new schema definition will try to update the current schema and type definitions. The result of this is either an error or `hotRod.Type` should inherit from `CarModel`.

Basically, any update to type system from .NET code host is not obvious and should not break the underlying model.

**Comment by k-rus** [comment]
> Currently, Db.TypeOf<Car> and Db.TypeOf("Car") return the same object. That might actually be a bit confusing. In the former, it's clear that we want to go from a class to it's corresponding type object. In the later - do we want to get the type of the type named "Car", or the "Car" type itself?

I agree it is very confusing. I personally don't understand the meaning of using generics in the first case, i.e., `Db.TypeOf<Car>`. Is it a work around that the type cannot be given as argument?

In my opinion name `Typeof` is not suitable for us, since it collides with C# operator `typeof`, which is designed for single type layer and has two semantics: return type of given object or return type itself.

**Comment by per-samuelsson** [comment]
> I personally don't understand the meaning of using generics in the first case, i.e., Db.TypeOf<Car>. Is it a work around that the type cannot be given as argument?

You prefer ```Db.TypeOf(Type t)``` instead, i.e. ```Db.TypeOf(typeof(Car))``` to go from class to type object?

> In my opinion name Typeof is not suitable for us, since it collides with C# operator typeof, which is designed for single type layer and has two semantics: return type of given object or return type itself.

I don't agree. I think they resemble each other. Our ```TypeOf``` gives the type object of the given class; their ```typeof``` operator gives the metadata type object of a given class. Not sure what you mean that ```typeof```has two semantics, I know of only one.

**Comment by k-rus** [comment]
> I personally don't understand the meaning of using generics in the first case, i.e., Db.TypeOf. Is it a work around that the type cannot be given as argument?
> You prefer Db.TypeOf(Type t) instead, i.e. Db.TypeOf(typeof(Car)) to go from class to type object?
> 
> In my opinion name Typeof is not suitable for us, since it collides with C# operator typeof, which is designed for single type layer and has two semantics: return type of given object or return type itself.
> I don't agree. I think they resemble each other. Our TypeOf gives the type object of the given class; their typeof operator gives the metadata type object of a given class. Not sure what you mean that typeofhas two semantics, I know of only one.

Sorry, it was my mistake. Operator `typeof` is applicable only on type and returns to type instance. Thus I agree that `Db.Typeof<Car>().Name == "Car".`

**Comment by per-samuelsson** [comment]
After yesterdays discussions, we have been able to formalize quite a lot on how the type object set will look and react when generated from a given VMDBMS code model.

Here's a new code model I'm planning to use as a milestone, and build regression tests around. This model includes quite a lot of real-world challenges, including multi-level defined types, inheritance of both types and non-types, and "skipping" of intermediate levels in a define type hierarchy (e.g. a ```ScandinavianModel``` is missing between ```EuropeanModel``` and ```SwedishModel```).

```C#
class CarFabric {}
class EuropeanFabric : CarFabric {}

class CarModel {
  [Type] CarFabric Fabric;
}
class EuropeanModel : CarModel {
  [Type] EuropeanFabric Fabric;
}
class SwedishModel : EuropeanModel {}

class Car {
  [Type] CarModel Model;
}
class EuropeanCar : Car {
  [Type] EuropeanModel Model;
}
class ScandinavianCar : EuropeanCar {}
class SwedishCar : ScandinavianCar {
  [Type] SwedishModel Model;
}
class AsianCar : Car {}
```

**Comment by per-samuelsson** [comment]
Here is a second model I'm planning to introduce. This one focus on the effect of the matrix when the type is not specified until we go deeper into a class hierarchy (i.e. ```Device``` has no explicit type specified, while ```Phone``` do).

```C#
class PhoneModel {}

class Device {}
class Phone : Device {
  [Type] PhoneModel Model;
}
class SmartPhone : Phone {}
class Tablet : Device {}
```

**Comment by per-samuelsson** [comment]
Yet another interesting one:

```C#
class CarModel {}
class SUV : CarModel {}
class Car { [Type] SUV Model; }
```

Here the type model is more specific than the class using it.

**Comment by per-samuelsson** [comment]
In parallel to working with a way to implement the multi-level binding schema (and apply all the rules and constraints defined by the type model), I have started producing reference documentation in a new [Under the hood: the VMDBMS dynamic type system](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system) wiki page.

**Comment by k-rus** [comment]
I started to read the specification and I have one comment:
Inheritance relationship is not 1-1, it is 1-many, since one type instance can be inherited by several. For example:
```c#
class Car { }
class Volvo : Car { }
class Saab : Car { }
```
In this example `Car` is extended by two classes.

I confirm that the type relationship is 1-1.

**Comment by k-rus** [comment]
I think naming columns in the matrix with class names is incorrect approach and will lead to misunderstanding. If more than one class is defined on the same level then the column should be named by both class names. I suggest to either omit names as it was on the white board or call them like `Instance`, `Type`, `TypeType`, etc.

**Comment by per-samuelsson** [comment]
> Inheritance relationship is not 1-1, it is 1-many, since one type instance can be inherited by several.

Absolutely right. Good find. Thanks! I will adjust.

> I think naming columns in the matrix with class names is incorrect approach and will lead to misunderstanding. If more than one class is defined on the same level then the column should be named by both class names.

I dont agree. I think its fine. I thought like you did at first, just considered the Opel deriving Car. However, I think it's fine because what is there is the *most common denominator*, i.e. the shared base class we force instance classes sharing type to have. And it adds information that makes the whole binding more clear IMO.

**Comment by k-rus** [comment]
> I dont agree. I think its fine. I thought like you did at first, just considered the Opel deriving Car. However, I think it's fine because what is there is the most common denominator, i.e. the shared base class we force instance classes sharing type to have. And it adds information that makes the whole binding more clear IMO.

Which class name should be used for the instance column in the following case?
```c#
class Foo {};
class Bar {}
```
Should it be `Foo` or `Bar` or both?

I guess you want to present the class of the instances. I think the class name should be located in each cell instead of column and denoted in some way, e.g., bold or italic.

**Comment by per-samuelsson** [comment]
> Which class name should be used for the instance column in the following case?

The idea is that each matrix defines only the types for either a single instance class, or a hierarchy of database classes, so your example would define two matixes.

WTBS, sure, we can change the syntax if you think its important. My idea was instead that the actual type of an instance was denoted with the ```[Key:TypeKey]``` notation, like I showed in one of the samples.

> I guess you want to present the class of the instances. I think the class name should be located in each cell instead of column and denoted in some way, e.g., bold or italic.

I considered this initially, but I was afraid it would bloat the matrix and make it cluttery. But maybe I'm wrong.

**Comment by k-rus** [comment]
> The idea is that each matrix defines only the types for either a single instance class, or a hierarchy of such classes, so your example would define two, separate matixes.

Thank you for the clarification. It was not clear from the wiki page.



> I guess you want to present the class of the instances. I think the class name should be located in each cell instead of column and denoted in some way, e.g., bold or italic.
> I considered this initially, but I was afraid it would bloat the matrix and make it cluttery. But maybe I'm wrong.

Here the example, which can bring the confusion:
In the first matrix example, instance **7:1** will be presented by instance of class `Car`.  If there is an instance in the second row, e.g., **8:4**, its corresponding C# instance will be of class `Baz`, which inherits from `Car`. In my opinion specifying class `Car` and not specifying class `Baz` is misleading. Either correct closest classes should be specified or they should be completed omitted. The first is preferable, since the page describe mapping between type system and .NET code host.

**Comment by per-samuelsson** [comment]
> In the first matrix example, instance 7:1 will be presented by instance of class Car. If there is an instance in the second row, e.g., 8:4, its corresponding C# instance will be of class Baz, which inherits from Car.

You are mixing ```Car``` and ```Baz``` here. They are never part of the same section. I guess you mean ```Foo```, not ```Car```.

If I'm right in the above, what I wrote earlier is not at all violated:

> [...] it's fine because what is there is the *most common denominator*, i.e. the shared base class we force instance classes sharing type to have.

And Baz ```is``` Foo, in C# lingo.

I also explicitly clarify it in the legend:

> The "Baz" 4 cell indicates a type with that name and key. Because it is under the "Foo" 1 object, it reference the "Foo"-object as its base type, i.e. **"Baz" inherits "Foo"**.

**Comment by per-samuelsson** [comment]
@k-rus 

Since you are so generous with feedback, I would really appreciate a discussion regarding [1.10](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#110-single-class-using-deeper-type-hierarchy). This is a case that have quite a large impact on the implementation, and highly dictates how the upcoming Heads model can be designed.

**Comment by k-rus** [comment]
> In the first matrix example, instance 7:1 will be presented by instance of class Car. If there is an instance in the second row, e.g., 8:4, its corresponding C# instance will be of class Baz, which inherits from Car.
> You are mixing Car and Baz here. They are never part of the same section. I guess you mean Foo, not Car.

Sorry, you are correct of course.



> And Baz is Foo, in C# lingo.

correct, but 8:1 `is Baz` in addition and this information, I think, might be important in some discussion further down.



> Since you are so generous with feedback, I would really appreciate a discussion regarding 1.10. This is a case that have quite a large impact on the implementation, and highly dictates how the upcoming Heads model can be designed.

Since `Model` is defined as a database class, i.e., `[Database] public class Model { }`, then the types should be created as in the matrix. The cell `??` will be filled with auto-generated type instance. In practice it means bad data model design in the code first approach.

I am looking to other examples now and will post in this issue.

**Comment by k-rus** [comment]
The example [1.10](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#110-single-class-using-deeper-type-hierarchy) and [1.4](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#14-single-class-with-explicit-subclassed-type-class) looks to be the same. Therefore the same my comment is applicable for [1.4](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#14-single-class-with-explicit-subclassed-type-class), i.e., `??` is an auto-generated type.

In [1.5](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#15-single-class-with-explicit-baseclass-type-class) `??` is an auto-generated type instance. It can be named later if necessary as any instance.

In [1.6](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#16-subclassed-class-with-explicit-type-class) tuple/instance `5` named `"Opel_type"` is auto type, thus the name, imho, should be `"Opel_autotype"`.

[1.7](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#17-single-class-with-explicit-type-class-inheriting-instance-class) breaks the conceptual type model and should not be accepted, i.e., Starcounter will fail these classes.

Discussions in [1.8](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#18-explicit-type-class-introduced-in-subclass)

>Discussion: this can produce a case that is probably unlikely, but one we must know how to deal with. What if Car and CarModel is defined in one application. That loads first. Then an application with the Opel definition load. Now CarModel has gone from an instance class to a type class. This will also widen the type matrix of Car, that is already running.

We either always fail such case or try to upgrade the schema. If there are instances of `CarModel`, then the main strategy will be to fail them, otherwise all the instances have to be fixed, which is going to be expensive.

>Another variant would be Car was in a separate application, Opel and CarModel in the same. If Car loads first, it will neccessarily need to migrate when Opel and CarModel loads.

I don't see any problem with this scenario. The only upgrade fix will be needed is to create `"Car_autotypetype" 3`.

**Comment by per-samuelsson** [comment]
> The example 1.10 and 1.4 looks to be the same.

Absolutly right. Thanks for noting this.

**Comment by per-samuelsson** [comment]
Regarding 1.10 and your comment

> Since Model is defined as a database class, i.e., [Database] public class Model { }, then the types should be created as in the matrix. The cell ?? will be filled with auto-generated type instance. In practice it means bad data model design in the code first approach,

and also considering 

> 1.7 breaks the conceptual type model and should not be accepted, i.e., Starcounter will fail these classes.

it indicates there must be a redesign in the Heads-migration. I'll raise the issue at some time in the office so we can discuss it. I think I need the advice of @Starcounter-Jack too.

I'll get back to other feedback eventually.

**Comment by k-rus** [comment]
> it indicates there must be a redesign in the Heads-migration. I'll raise the issue at some time in the office so we can discuss it. I think I need the advice of @Starcounter-Jack too.

I think we can introduce additional auto-types to break the circle of inheritance and type relationship. A possible implementation for 1.7 can be:

```C#
class Category : Root {}
class Root {
  [Type] Category Category;
}
```

Root | Category | object | object
-----|-----|-----|------
 |Root|auto|??
 |auto|Category|??

This proposal is for discussion.

**Comment by per-samuelsson** [comment]
> A possible implementation for 1.7 can be:

What is important, if we need to preserve the proposed Heads migration model, is that Root (and subclasses) act in a way that they only instantiate. I guess we need to discuss it, yes. To make sure no flaw is present, and if we can model it in some alternative way if needed.

**Comment by per-samuelsson** [comment]
> In 1.6 tuple/instance 5 named "Opel_type" is auto type, thus the name, imho, should be "Opel_autotype". I guess it is just a typo.

Fixed.

> In 1.5 ?? is an auto-generated type instance. It can be named later if necessary as any instance.

Regarding this, and my more recent tweak in [1.11](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood:-the-VMDBMS-dynamic-type-system#111-single-class-with-explicit-baseclass-type-class-subtype-being-broader), I'm thinking it might be an option just to ignore the more specialized type model (i.e. ```SUV```) as a whole, since it's not in use in the sense that no instance class ever define it?

Or, if we decide to stick with it, I guess the actual code host type we instantiate for any new Car should be ```SUV```, not ```CarModel```? That is, the following should be **true**: ```new Car().Type.GetType() == typeof(SUV)```?

A third alternative might be to just not support it at all (at least to begin with)? If so, the declaration of ```Car``` would not be allowed since ```CarModel``` is not a leaf class (i.e. we force ```Car``` to define ```SUV```).

Any opinions?

**Comment by k-rus** [comment]
> In 1.5 ?? is an auto-generated type instance. It can be named later if necessary as any instance.
> Regarding this, and my more recent tweak in 1.11, I'm thinking it might be an option just to ignore the more specialized type model (i.e. SUV) as a whole, since it's not in use in the sense that no instance class ever define it?
> 
> Or, if we decide to stick with it, I guess the actual code host type we instantiate for any new Car should be SUV, not CarModel? That is, the following should be true: new Car().Type.GetType() == typeof(SUV)?
> 
> A third alternative might be to just not support it at all (at least to begin with)? If so, the declaration of Car would not be allowed since CarModel is not a leaf class (i.e. we force Car to define SUV).
> 
> Any opinions?

It is not necessary to be leaf node to create instances. So in the model of 1.5 it is possible to create non-type instances from `Car` and from the instance of `SUV`, which inherits `Car` and which can be retrieved using SQL. I don't see any need to prevent the situation.

**Comment by per-samuelsson** [comment]
Just to peak into what we have in front of us to implement, after we have fully solved static analysis, I want to just briefly discuss the ```New()``` implementation applied on a rather complex model.

Considering [1.100 model](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood;-the-VMDBMS-dynamic-type-system#1100-complex-milestone), can we try pseudo-code the operation of the ```New()``` method when invoked from the 201-type?

Now, I'm a correct that it should produce **8 new type objects**? Initially one new that has 201 as its parent (i.e. ```Inherits=201```), and then we need a type of that, and then "fill out" all layers "below"?

I guess some kind of smart indexing (likely including type ID's, when they are in place) is needed, but just considering pseudo-code, how is it solvable? 

Something like this?

1. Create a new type, say 350. Set inherits to 201.
2. Create a new type, 351. Set 350.Type=351. Set 351.Instantiates=350.
3. Look for 201.Type. If it exist, set 351.Inherits=201.Type. Otherwise, create 201.Type and do some recursive thing similar to what we do in 4-x below.
4. Do a select to find all objects "in the same cell" as our new 350, i.e. all objects that directly derive 201. Go to the "left" from any such object, to create types. Go to the "right" from such object, to create new types.
5. Find all direct subtypes of every type below, recursively, and do something similar to [4].

**Comment by k-rus** [comment]
> Look for 201.Type. If it exist, set 351.Inherits=201.Type. Otherwise, create 201.Type and do some recursive thing similar to what we do in 4-x below.

Am I correct that it means that 351.Inherits = 203?

In the step 4 I would create an object in any case due to definition of `New`. So I will create 352 and set 352.Type = 350. Then I find that 352 should inherit from 200, since 200.Type = 201. Then since 200.Instance = AsianCar I will create 353 and set 353.Inherits AsianCar and 353.Type = 352.

Then I would stop. This I guess is more simplistic than implemented. I don't remember why it is necessary to go down. Otherwise, it requires to create 4 more objects as suggested.

**Comment by per-samuelsson** [comment]
> Am I correct that it means that 351.Inherits = 203?

Correct, since in the sample, 201 actually had a type already (i.e. 203). My comment was more general.

**Comment by per-samuelsson** [comment]
This has been reevaluated, and will be put in the backlog. Part of the functionality implemented has been cherrypicked and merged as part of #2630.
